File: sleepermetrics/R/players.R
# Player metadata (league-independent, cached) ----------------------------

.sl_player_cache <- new.env(parent = emptyenv())

#' Player name / position lookup
#'
#' Downloads Sleeper's full NFL player database (the largest payload it serves)
#' and reduces it to a `player_id -> name/position/gsis_id` table. The dump is
#' league-independent, so it is cached on disk (refetched at most once per day)
#' and memoised for the session.
#'
#' `DEF` "players" are team defenses; their `player_name` is set to the team
#' abbreviation (the Sleeper `player_id`).
#'
#' @param refresh Force a re-download even if today's cache exists.
#' @param cache_path RDS cache location.
#' @return A tibble with `player_id`, `player_name`, `position`, `gsis_id`.
#' @export
sl_players <- function(refresh = FALSE,
                       cache_path = "sleeperPlayerData.rds") {
  key <- "info"
  if (!refresh && !is.null(.sl_player_cache[[key]])) return(.sl_player_cache[[key]])
  fresh <- file.exists(cache_path) &&
    as.Date(file.info(cache_path)$mtime) == Sys.Date()
  raw <- if (fresh) readRDS(cache_path) else {
    r <- sleeper_api("/players/nfl"); saveRDS(r, cache_path); r
  }
  clean <- function(x, empty = NA) {
    if (is.null(x)) return(NA)
    if (is.list(x)) {
      if (!length(x)) return(empty) else return(lapply(x, clean, empty = empty))
    }
    x
  }
  df <- purrr::map_dfr(clean(unname(raw)),
                       ~ tibble::as_tibble(jsonlite::flatten(as.data.frame(.x))))
  info <- df %>%
    dplyr::transmute(
      player_id,
      player_name = dplyr::if_else(position == "DEF",
                                   as.character(player_id), full_name),
      position,
      gsis_id = as.character(gsis_id))
  .sl_player_cache[[key]] <- info
  info
}

File: python/sleepermetrics/players.py
"""Player metadata lookup, cached daily (mirrors R players.R)."""
from __future__ import annotations

import datetime
import os
import pickle

import pandas as pd

from .api import sleeper_api

_cache: pd.DataFrame | None = None


def players(refresh: bool = False, cache_path: str = "sleeperPlayerData_py.pkl") -> pd.DataFrame:
    """Return a DataFrame: player_id, player_name, position, gsis_id.

    DEF "players" are team defenses; their name is the team abbreviation.
    The full player dump is cached on disk (refetched at most once/day) and
    memoised for the process.
    """
    global _cache
    if _cache is not None and not refresh:
        return _cache
    fresh = (
        os.path.exists(cache_path)
        and datetime.date.fromtimestamp(os.path.getmtime(cache_path)) == datetime.date.today()
    )
    if fresh and not refresh:
        with open(cache_path, "rb") as fh:
            raw = pickle.load(fh)
    else:
        raw = sleeper_api("/players/nfl")
        with open(cache_path, "wb") as fh:
            pickle.dump(raw, fh)
    rows = []
    for pid, p in raw.items():
        if not isinstance(p, dict):
            continue
        pos = p.get("position")
        rows.append({
            "player_id": pid,
            "player_name": pid if pos == "DEF" else p.get("full_name"),
            "position": pos,
            "gsis_id": p.get("gsis_id"),
        })
    _cache = pd.DataFrame(rows)
    return _cache
